### PR TITLE
App for signed random data

### PR DESCRIPTION
**Pull request by dehanj**
New features:
- Good generation of random data (form rng_strem)
- hashes the data and signes it to be able to verify the origin of the data
- Can output the data to a binary file
- Validates hashes and signature in client application

Should probably be moved to its own repo in the future, so we can make releases and distribute via brew etc.

### CODE REVIEW

File: cmd/runrandom/proto.go [hunk 1]
```diff
@@ -138,3 +144,74 @@ func (s RandomGen) GetRandom(bytes int) ([]byte, error) {
...
+}
+
+func (s RandomGen) GetHash() ([]byte, error) {
+	id := 2
+	tx, err := tkeyclient.NewFrameBuf(cmdGetHash, id)
+	if err != nil {
+		return nil, fmt.Errorf("NewFrameBuf: %w", err)
+	}
+
+	tkeyclient.Dump("GeHash tx", tx)
```
Comment (secworks) [inline cmd/runrandom/proto.go:200]: Nits: Spelling "GeHash"

**Review by secworks — COMMENTED** [review]
Looks good. Only saw a nits. The code does what I expected. One could possibly use isatty to better control where to send data (data, pubkey, signature) - something I realised after discussing solutions.

File: cmd/runrandom/proto.go (hunk 1 above)
Comment (dehanj) [reply cmd/runrandom/proto.go:200]: Thanks, fixed in the new repo.

**Comment by dehanj** [comment]
There is now a new repository made for this application: https://github.com/tillitis/tkey-random-generator. It includes both client and device application. 

All comments/changes requested here will be made in that repo. Then this PR will be closed without merging.
I will let this PR live until we agree that it looks ok.

File: apps/random/rng.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+uint8_t rng_initalized = 0;
+
+static uint32_t entropy_get()
+{
+	while ((*trng_status & (1 << TK1_MMIO_TRNG_STATUS_READY_BIT)) == 0) {
+	}
+	return *trng_entropy;
+}
+
+// Do you really keep the entropy for byte 32-63?
```
Comment (mchack-work) [inline apps/random/rng.c:27]: What does this mean?

File: apps/random/rng.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	}
+
+	ctx->ctr = 0;
+
+	rng_initalized = 1;
+}
+
+int rng_get(uint32_t *output, rng_ctx *ctx, int size)
+{
+
```
Comment (mchack-work) [inline apps/random/rng.c:62]: Unnecessary whitespace

File: apps/random/main.c [hunk 4]
```diff
@@ -50,10 +34,31 @@ int main(void)
 	uint8_t rsp[CMDLEN_MAXBYTES];
 	uint8_t in;
```
Comment (mchack-work) [inline apps/random/main.c:36]: Why a newline?

File: apps/random/main.c [hunk 5]
```diff
@@ -50,10 +34,31 @@ int main(void)
...
 
+	uint32_t digest[32];
+	uint8_t pubkey[32];
+	uint32_t local_cdi[8];
+	uint8_t signature[64];
+	uint8_t hash[32];
+	uint8_t signature_generated = 0;
+	uint8_t rand_data_generated = 0;
+
```
Comment (mchack-work) [inline apps/random/main.c:44]: Why newline?

File: apps/random/README.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+## Application protocol
+
+`Random generator` has a simple protocol on top of the [TKey Framing
+Protocol](https://dev.tillitis.se/protocol/#framing-protocol) with the
+following requests:
+
+| *command*             | *FP length* | *code* | *data*                              | *response*            |
+|-----------------------|-------------|--------|-------------------------------------|-----------------------|
+| `CMD_GET_NAMEVERSION` | 1 B         | 0x01   | none                                | `RSP_GET_NAMEVERSION` |
+| `CMD_GET_RANDOM`      | 4 B         | 0x03   | Number of bytes, 1 < x < 126        | `RSP_GET_RANDOM`      |
```
Comment (mchack-work) [inline apps/random/README.md:21]: Why ask for just 126 bytes at a time? This leads to a lot of stuttering, basically Xmodem style. Another idea would be to ask for N bytes here, perhaps a u32, and then reply with many frames until N bytes (padded) has been sent back.

I'm OK with this at is is. Just wondering if it would lead to better performance. Yes, I think we wrote something about always doing ping-pong on the framing level... Nevertheless.

**Review by mchack-work — CHANGES_REQUESTED** [review]

File: apps/random/main.c [hunk 7]
```diff
@@ -107,17 +112,72 @@ int main(void)
...
+		case APP_CMD_GET_PUBKEY:
+			qemu_puts("APP_CMD_GET_PUBKEY\n");
+			memcpy(rsp, pubkey, 32);
+			appreply(hdr, APP_RSP_GET_PUBKEY, rsp);
+			break;
+
+		case APP_CMD_GET_HASH:
+			qemu_puts("APP_CMD_GET_HASH\n");
+			if (signature_generated == 0) {
+				qemu_puts("Requested hash before siganture is "
```
[the remark this answers is not in the document]
Comment (dehanj) [reply apps/random/main.c:146]: Fixed in new branch.

File: cmd/runrandom/main.go [hunk 8]
```diff
@@ -121,18 +142,87 @@ Usage:
...
+			// Don't exit, let's calculate hash
+		} else {
+			le.Printf("signature verified.\n")
+		}
+
+		localHash := blake2s.Sum256(totRandom)
+
+		le.Printf("\nVerifying hash ... ")
+
+		if !testEq(hash, localHash) {
```
[the remark this answers is not in the document]
Comment (dehanj) [reply cmd/runrandom/main.go:202]: Ah, that explains it! I tried to use it, but GO refused to add the bytes package, which is due to the conflict of the local variable - which I for some reason missed. Anyhow, will use `bytes.Equal() instead`. Updated in new repo.

File: cmd/runrandom/main.go [hunk 9]
```diff
@@ -121,18 +142,87 @@ Usage:
...
+		}
+		le.Printf("hash verified.\n")
+
+	}
+
+	file.Close()
 	exit(0)
 }
 
+func testEq(a []byte, b [32]byte) bool {
```
[the remark this answers is not in the document]
Comment (dehanj) [reply cmd/runrandom/main.go:214]: Done in new repo.

File: apps/random/README.md [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+The TKey `Random generator` device application is an hardware based
+source of high quality random numbers. The generator is using a
+Hash_DRBG built around the BLAKE2s hash function. The application can
+also sign the random data in order to provide proof of its origin.
+
+## Client Go application
+
+The corresponding client application can be found in `cmd/runrandom`.
+
+## Application protocol
```
[the remark this answers is not in the document]
Comment (dehanj) [reply apps/random/README.md:12]: The two READMEs are now unified in the top-level in the new repo.

File: apps/random/rng.c (hunk 2 above)
Comment (dehanj) [reply apps/random/rng.c:27]: This was a memory note to discuss with Joachim. It is taken care of so I'm removing it in the new branch.

File: cmd/runrandom/main.go [hunk 11]
```diff
@@ -121,18 +142,87 @@ Usage:
...
+		} else {
+			le.Printf("signature verified.\n")
+		}
+
+		localHash := blake2s.Sum256(totRandom)
+
+		le.Printf("\nVerifying hash ... ")
+
+		if !testEq(hash, localHash) {
+			le.Printf("hash FAILED verification.\n")
```
[the remark this answers is not in the document]
Comment (dehanj) [reply cmd/runrandom/main.go:203]: After our discussion, I will create an issue about this and leave as is.

File: apps/random/rng.c (hunk 3 above)
Comment (dehanj) [reply apps/random/rng.c:62]: Done.

Comment (dehanj) [reply apps/random/rng.c:62]: Done.

File: apps/random/main.c (hunk 4 above)
Comment (dehanj) [reply apps/random/main.c:36]: Think it increases readability to group them. But fixed in new repo.

File: apps/random/main.c (hunk 5 above)
Comment (dehanj) [reply apps/random/main.c:44]: Fixed in new repo.

File: apps/random/README.md (hunk 6 above)
Comment (dehanj) [reply apps/random/README.md:21]: Agreeing on the performance increase by removing the ping-pong. Will create an issue about this.

**Comment by dehanj** [comment]
This application now lives at: https://github.com/tillitis/tkey-random-generator
